book/src/generalized-ownership.md

```diff
@@ -0,0 +1,284 @@
+
+Alice requests X amount of token implemented by program Token1, in
+exchange for Y amount of token implemented by program Token2.
+
+Bob offers X + 1 amount of token implemented by program Token1, in exchange for
+Y - 1 amount of token implemented by Token2.
+
+Dan pairs Bob and Alice, causing the resulting ownership changes:
+
+* Alice owns X of Token1,
+* Bob owns Y - 1 of Token2,
+* Dan owns 1 of Token1 and 1 of Token2,
+
+Another example:
+
+Alice offers X amount of Token1 to the winner of TicTacToe.
+
+Bob offers Y amount of Token2 to the winner of the same game of TicTacToe.
+
+The game charges a commission, the ceiling of a percentage of the tokens.  The
+winner takes ownership of both tokens minus the commission.  The commission is
+transferred to the address designated by the game.
+
+## Current System and Token programs
+
+The System program defines ownership of lamports to be identified by the public key
+that matches the account's pubkey. The`SystemInstruction::Transfer` instruction checks that
+the identity of the spending account has a valid signature before completing the
+transfer.
+
+The Token program defines ownership of the token to be identified by the public key
+that matches the account's pubkey, and `Token::Transfer` instruction checks
+that the identity of the spending account has a valid signature before
+completing the transfer.
+
+Both of these programs implement almost the same mechanism.
+ 
+## General idea
+
+For items to have externally countable ownership, they must rely on the external
+program for managing the transfer of units.  These items must have a valid state
+where the number of units owned is 0.  The **zero** count of items is
+the destination for transferring any count of items on the chain.
+
+Execution of these state transitions must allow for a reentrant and partial
+transaction to make progress.
+
+## Generalized owner identity
+
+Using the account's pubkey identity as the asset owner identity means that all
+the asset transfer APIs must be re-implemented by the program code.
+
+The proposed solution is to use a pointer to an owner identity.
+
+```rust,ignore
+// This state is owned by the Owner program
+enum OwnerState {
+    // the owner identity must provide a valid signature for this pubkey
+    Unique {
+        // The actual owner key.
+        owner: Pubkey, 
+        // The item program instance.
+        item: Pubkey
+    },
+    //...
+}
+```
+
+* OwnerState::Unique.owner - The actual OwnerState’s account’s pubkey is not used to
+distinguish ownership.  Only the `OwnerState::Unique.owner` pubkey identifies the
+ownership identity.
+
+```rust,ignore
+// This state is owned by the MyUniqueTokenState
+struct MyUniqueTokenState {
+    // This points to an OwnerState structure that identifies the actual owner
+    owner: Pubkey,
+}
+```
+
+* MyUniqueStokenState.owner - address of the OwnerState that identifies the
+current owner.
+
+### OwnerState::Transfer
```

Comment:
Something with a the suffix "State" has typically been the struct whose instances are serialized into `Account.data`s.  Something with the suffix "Instruction" has typically been an enum whose instances are serialized into `Instruction.data`s.  I don't understand what is account state and what is instruction data in this proposal. These code comments pointing to accounts suggest these are instructions, not state.